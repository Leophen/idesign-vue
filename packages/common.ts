import _ from 'lodash'
import {
  Component,
  getCurrentInstance,
  Slots,
  VNode,
} from 'vue'

// 通用位置类型
export type placementType =
  | 'top'
  | 'top-left'
  | 'top-right'
  | 'bottom'
  | 'bottom-left'
  | 'bottom-right'
  | 'left'
  | 'left-top'
  | 'left-bottom'
  | 'right'
  | 'right-top'
  | 'right-bottom'

// 通用触发类型
export type triggerType = 'hover' | 'click' | 'context-menu'

// 转换 width/height/left/... 样式值
export const turnValue = (val: string | number | undefined) => {
  if (!val) {
    return
  }
  if (_.isNumber(val as number)) {
    return `${val}px`
  }
  return val
}

// const _stateValue = ref(defaultValue)
// const innerValue = computed(() => value ?? _stateValue.value)

/**
 * 渲染 default slot，获取子组件 VNode。处理多种子组件创建场景
 * @returns {function(childComponentName: string, slots: Slots): VNode[]}
 * @param childComponentName
 * @param slots
 * @example const getChildByName = useChildComponentSlots()
 * @example getChildComponentByName('xxItem')
 */
export const useChildComponentSlots = () => {
  const instance = getCurrentInstance()
  return (childComponentName: string, slots?: Slots): VNode[] => {
    if (!slots) {
      slots = instance?.slots
    }
    const content = slots?.default?.() || []

    return (
      content
        .map((item: VNode) => {
          if (item.children && Array.isArray(item.children))
            return item.children
          return item
        })
        .flat()
        // @ts-ignore
        .filter((item: VNode) =>
          (item.type as Component).name?.endsWith(childComponentName)
        ) as VNode[]
    )
  }
}
