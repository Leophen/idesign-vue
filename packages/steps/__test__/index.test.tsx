import { Steps, StepsItem } from '../index';
import { render } from '../../common'

describe('Steps 组件测试', () => {
  it('create', () => {
    const wrapper = render(
      <Steps>
        <StepsItem
          v-slots={{
            title: () => '步骤1',
            description: () => '提示文字1'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤2',
            description: () => '提示文字2'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤3',
            description: () => '提示文字3'
          }}
        />
      </Steps>
    );
    expect(wrapper.element).toMatchSnapshot();
  });

  it('current', () => {
    const wrapper = render(
      <Steps current={2}>
        <StepsItem
          v-slots={{
            title: () => '步骤1',
            description: () => '提示文字1'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤2',
            description: () => '提示文字2'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤3',
            description: () => '提示文字3'
          }}
        />
      </Steps>
    );
    expect(wrapper.findAll('.i-steps-item')[0].classes()).toContain('i-steps-item__finish');
    expect(wrapper.findAll('.i-steps-item')[1].classes()).toContain('i-steps-item__current');
  });

  it('dot', () => {
    const wrapper = render(
      <Steps dot>
        <StepsItem
          v-slots={{
            title: () => '步骤1',
            description: () => '提示文字1'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤2',
            description: () => '提示文字2'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤3',
            description: () => '提示文字3'
          }}
        />
      </Steps>
    );
    expect(wrapper.classes()).toContain('i-steps__dot');
  });

  it('layout', () => {
    const wrapper = render(
      <Steps layout="vertical">
        <StepsItem
          v-slots={{
            title: () => '步骤1',
            description: () => '提示文字1'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤2',
            description: () => '提示文字2'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤3',
            description: () => '提示文字3'
          }}
        />
      </Steps>
    );
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.classes()).toContain('i-steps__layout-vertical');
  });

  it('reverse', () => {
    const wrapper = render(
      <Steps reverse>
        <StepsItem
          v-slots={{
            title: () => '步骤1',
            description: () => '提示文字1'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤2',
            description: () => '提示文字2'
          }}
        />
        <StepsItem
          v-slots={{
            title: () => '步骤3',
            description: () => '提示文字3'
          }}
        />
      </Steps>
    );
    expect(wrapper.element).toMatchSnapshot();
    expect(wrapper.findAll('.i-steps-item')[0].find('.i-steps-item__description').text()).toBe('提示文字3');
  });
});

