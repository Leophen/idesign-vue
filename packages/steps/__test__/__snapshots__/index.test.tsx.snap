// Vitest Snapshot v1

exports[`Steps 组件测试 > create 1`] = `
<div
  class="i-steps i-steps__layout-horizontal"
>
  
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤1
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            1
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字1
        
      </article>
    </div>
  </div>
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤2
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            2
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字2
        
      </article>
    </div>
  </div>
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤3
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            3
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字3
        
      </article>
    </div>
  </div>
  
</div>
`;

exports[`Steps 组件测试 > layout 1`] = `
<div
  class="i-steps i-steps__layout-vertical"
>
  
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤1
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            1
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字1
        
      </article>
    </div>
  </div>
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤2
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            2
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字2
        
      </article>
    </div>
  </div>
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤3
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            3
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字3
        
      </article>
    </div>
  </div>
  
</div>
`;

exports[`Steps 组件测试 > reverse 1`] = `
<div
  class="i-steps i-steps__layout-horizontal"
>
  
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤3
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            3
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字3
        
      </article>
    </div>
  </div>
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤2
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            2
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字2
        
      </article>
    </div>
  </div>
  <div
    class="i-steps-item"
  >
    <div
      class="i-steps-item__content"
    >
      <header
        class="i-steps-item__title"
      >
        
        步骤1
        
        <div
          class="i-steps-item__icon"
        >
          <span>
            1
          </span>
        </div>
      </header>
      <article
        class="i-steps-item__description"
      >
        
        提示文字1
        
      </article>
    </div>
  </div>
  
</div>
`;
